Terminal.Gui/Views/Scroll.cs

```diff
@@ -60,7 +60,7 @@ public int Position
         get => _position;
         set
         {
-            int barSize = Orientation == Orientation.Vertical ? ContentSize.Height : ContentSize.Width;
+            int barSize = Orientation == Orientation.Vertical ? GetContentSize ().Height : GetContentSize ().Width;
```

Comment:
Why was this done all over the file? This is a concurrency problem and a performance issue resulting in more than 2x the work

Reply:
I'll have an alternative to propose to you in just a few minutes, for this.

Reply:
Also. What happens when `value == 0`?

The two checks in the if statement right below it have the same outcome (but ever so slightly lower performance than) this:

```csharp
switch (value)
{
    case < 0:
    case > 0 when value + barSize > Size:
        return;
}
```

So, value==0 will fall through, but require that two checks be performed, in the existing form (`value < 0 || (value > 0 && value + barSize > Size)`) because, when value is 0:
 1. `value < 0` is evaluated unconditionally no matter the value.
     * result is false, so right operand of the `||` is evaluated
 2. `value > 0` is then evaluated unconditionally.
     * Evaluates to false.
 3. The logical && operator function is called, but due to assembly/RPN magic, it short circuits and returns false before evaluating the final operand (as expected).
     * Result is false, and statement ends, passing control to line 70
 4. Line 70 unconditionally calls `OnPositionChanging (_position, value);` with value=0.
 5. Event is raised with that value being 0, which is why the unit test has to special case 0 and cancel.

This can all be avoided by just dealing with the 0 case. That is particularly important if the consumer has subscribed to the event and now we've given them a pointless event called PositionChanging that isn't actually a changing position, and then also have to have the workaround in both the unit tests AND the class to special case the 0.

Reply:
Remember:

When the event is raised, you do not have a guarantee of the order the members in the MulticastDelegate will be called. Their code could run first. Our code could run first. Their code could be async or parallel and run concurrently with our code. Who knows. We can't depend on that.

Raising the event for an invalid state should be avoided, not handled post-hoc.

Reply:
OK.

So...

That distraction has changed what my recommendation for addressing the original comment was going to be, but it's a bit more involved than is really appropriate to give code snippets for here.

The extra commentary may also change as a result of this, so consider this part first:

Unfortunately, the GetContentSize method creates the root problem (the property had the same problem, but not because it was a property).

So, those methods all need to know the size of the scroll bar, clearly. And that method is a perfectly reasonable idea for delegating the responsibility to a container, in various other situations.

But, even in those other situations, it is sub-optimal, for performance, on top of the local issues described above.

Much more optimally, it should store its own size.

"But how would it handle its container changing size?"

Events!

But the explanation is a lot more than the code itself, a significant amount of which actually does already exist. :)

I'm not sure right at the moment what my ultimate recommendation would be for a "proper" fix, here.

Perhaps just what the original recommendation was going to be, which would be changing the beginning of the set accessor for Position to be:

```csharp
            if (value <= 0)
            {
                return;
            }

            Size contentSize = GetContentSize ();
            int barSize = Orientation == Orientation.Vertical ? contentSize.Height : contentSize.Width;

            if (value + barSize > _size)
            {
                return;
            }
```

This does the following:
 * Covers the 0 case by returning immediately.
 * Also returns before bothering to get and assign `barSize`, which is at least 3 copies.
 * The conditional compiles to the same code as before, so that's an optional change.
 * Uses the field instead of the property in the last check.

This does not address the various other things I brought up, however.

But that can be done another time.

Reply:
> This does the following:
> 
> * Covers the 0 case by returning immediately.

I changed the condition to return if value equals position or is <0 or value + _barSize > Size.
Remember that I have to consider 0 as valid if the current position is different.

> * Also returns before bothering to get and assign `barSize`, which is at least 3 copies.

I created a private field `_barSize` and I think its value will be cached. On the other hand, it will only be executed if both previous conditions are not met.

> * The conditional compiles to the same code as before, so that's an optional change.
> * Uses the field instead of the property in the last check.
> 
> This does not address the various other things I brought up, however.
> 
> But that can be done another time.

I think it's more fluid now but I'm sure you would improve it enormously when you deal with this subject.

Reply:
Yep that's a better way for now.

I'll have a look at it in a minute to see if anything jumps out at me.

Reply:
I'm a bit unclear on what you mean about the 0 case. 0 would mean no position change. And the unit test explicitly cancels for that case anyway.

Reply:
No, 0 is position 0. If position is 3 and  then is defined as 0 it must be considered to invoke the events.

Reply:
Ah, ok, gotcha.

That wasn't clear to me last night, but I'll take another look at it later with that in mind.

Reply:
But do take note of the comment I added about the _position property (it's not a field). That wasn't quite what I was going for and just moved the same "problem" to a different place.

Reply:
I think you are talking about _barSize and it really isn't a field but a getter property. I only did what you asked for which was moving to only one place. It must be a calculated variable and the way I have it before was better because it's only called once on each place.

Reply:
Yep. I understand why you did it, and it's _almost_ what I was going for.

However, using a property or other type of method (remember a property accessor is a method) still leaves open the concurrency problem I mentioned, because it's entirely possible for the same method to ask for the property twice and get a different value for each access.

The methods in this class that use that new property do have multiple uses of it.

Storing the value in a local and using the local throughout the method is safer, and ensures that you will always have the same value at all uses in the scope of the method.

While you did in fact reduce some duplication, that wasn't the main goal - fixing the concurrency issue was the important part.

Reply:
Another thing is that making a property dependent upon another class directly, that way, creates a non-obvious and tight coupling between the actual class and the other class. And it being private hides what's actually going on from a consumer, which makes their type actually dependent on that property without knowing it.

Use of a local restricts the coupling to the scope of the specific methods, at least.

What I'd do, if you like the private member means of getting it, is make the property a method instead, call it once right before the first time it is used in those two methods, store the returned int in a local variable, and then use the variable only, after that, in the method.